go/store/datas/puller_test.go

```diff
@@ -0,0 +1,493 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package datas
+
+import (
+	"context"
+	"errors"
+	"os"
+	"path/filepath"
+	"testing"
+
+	"github.com/google/uuid"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+
+	"github.com/liquidata-inc/dolt/go/store/nbs"
+	"github.com/liquidata-inc/dolt/go/store/types"
+	"github.com/liquidata-inc/dolt/go/store/util/clienttest"
+)
+
+func mustTuple(tpl types.Tuple, err error) types.Tuple {
+	if err != nil {
+		panic(err)
+	}
+
+	return tpl
+}
+
+func addTableValues(ctx context.Context, vrw types.ValueReadWriter, m types.Map, tableName string, alternatingKeyVals ...types.Value) (types.Map, error) {
+	val, ok, err := m.MaybeGet(ctx, types.String(tableName))
+
+	if err != nil {
+		return types.EmptyMap, err
+	}
+
+	var tblMap types.Map
+	if ok {
+		mv, err := val.(types.Ref).TargetValue(ctx, vrw)
+
+		if err != nil {
+			return types.EmptyMap, err
+		}
+
+		me := mv.(types.Map).Edit()
+
+		for i := 0; i < len(alternatingKeyVals); i += 2 {
+			me.Set(alternatingKeyVals[i], alternatingKeyVals[i+1])
+		}
+
+		tblMap, err = me.Map(ctx)
+
+		if err != nil {
+			return types.EmptyMap, err
+		}
+	} else {
+		tblMap, err = types.NewMap(ctx, vrw, alternatingKeyVals...)
+
+		if err != nil {
+			return types.EmptyMap, err
+		}
+	}
+
+	tblRef, err := writeValAndGetRef(ctx, vrw, tblMap)
+
+	if err != nil {
+		return types.EmptyMap, err
+	}
+
+	me := m.Edit()
+	me.Set(types.String(tableName), tblRef)
+	return me.Map(ctx)
+}
+
+func deleteTableValues(ctx context.Context, vrw types.ValueReadWriter, m types.Map, tableName string, keys ...types.Value) (types.Map, error) {
+	if len(keys) == 0 {
+		return m, nil
+	}
+
+	val, ok, err := m.MaybeGet(ctx, types.String(tableName))
+
+	if err != nil {
+		return types.EmptyMap, err
+	}
+
+	if !ok {
+		return types.EmptyMap, errors.New("can't delete from table that wasn't created")
+	}
+
+	mv, err := val.(types.Ref).TargetValue(ctx, vrw)
+
+	if err != nil {
+		return types.EmptyMap, err
+	}
+
+	me := mv.(types.Map).Edit()
+	for _, k := range keys {
+		me.Remove(k)
+	}
+
+	tblMap, err := me.Map(ctx)
+
+	if err != nil {
+		return types.EmptyMap, err
+	}
+
+	tblRef, err := writeValAndGetRef(ctx, vrw, tblMap)
+
+	if err != nil {
+		return types.EmptyMap, err
+	}
+
+	me = m.Edit()
+	me.Set(types.String(tableName), tblRef)
+	return me.Map(ctx)
+}
+
+func tempDirDB(ctx context.Context) (Database, error) {
+	dir := filepath.Join(os.TempDir(), uuid.New().String())
```

Comment:
ioutil.TempDir is more appropriate for tests:

https://golang.org/pkg/io/ioutil/#example_TempDir